base/serialize.jl

```diff
@@ -419,27 +418,32 @@ function serialize_typename_body(s::AbstractSerializer, t::TypeName)
     else
         writetag(s.io, UNDEFREF_TAG)
     end
+    nothing
 end
 
 # decide whether to send all data for a type (instead of just its name)
-function should_send_whole_type(s, t::ANY)
+function should_send_whole_type(s, t::DataType)
     tn = t.name
     if isdefined(tn, :mt)
         # TODO improve somehow
         # send whole type for anonymous functions in Main
-        fname = tn.mt.name
+        name = tn.mt.name
         mod = tn.module
-        toplevel = isdefined(mod, fname) && isdefined(t, :instance) &&
-            getfield(mod, fname) === t.instance
-        ishidden = unsafe_load(unsafe_convert(Ptr{UInt8}, fname))==UInt8('#')
-        return mod === __deserialized_types__ || (mod === Main && (ishidden || !toplevel))
+        isanonfunction = mod === Main && # only Main
+            t.super === Function && # only Functions
+            unsafe_load(unsafe_convert(Ptr{UInt8}, tn.name)) == UInt8('#') && # hidden type
+            (!isdefined(mod, name) || t != typeof(getfield(mod, name))) # XXX: 95% accurate test for this being an inner function
```

Comment:
That seems suboptimal (the types won't be egal), although in that case it's still broken in the other direction (an inner function can be confused for an outer function in the case of code such as `f() = g() = 2; const g = f()`)

fwiw, getting this right would also make anonymous and inner method show printing more readable. for now, I simply added a TODO comment there for revisiting later so that blame will indicate both files)